File: Daily-Streak/Day-04_Four-Divisors/Solution.cpp
class Solution { 
public:
    int sumFourDivisors(vector<int>& nums) {
        int answer = 0;

        // ✅ OPTIMAL APPROACH (ACTIVE CODE)
        for (int x : nums) {
            int count = 0;
            int sum = 0;

            // Check divisors only till sqrt(x)
            for (int i = 1; i * i <= x; i++) {
                if (x % i == 0) {
                    int d1 = i;
                    int d2 = x / i;

                    count++;
                    sum += d1;

                    // If divisor pair is different, count both
                    if (d1 != d2) {
                        count++;
                        sum += d2;
                    }

                    // Early stopping if more than 4 divisors
                    if (count > 4) break;
                }
            }

            // If exactly 4 divisors, add their sum
            if (count == 4) {
                answer += sum;
            }
        }

        return answer;
    }

    /*
    ❌ BRUTE FORCE APPROACH (COMMENTED – FOR LEARNING ONLY)

    int sumFourDivisors(vector<int>& nums) {
        int ans = 0;
        for (int x : nums) {
            vector<int> divisors;
            for (int i = 1; i <= x; i++) {
                if (x % i == 0) {
                    divisors.push_back(i);
                }
            }
            if (divisors.size() == 4) {
                for (int d : divisors) ans += d;
            }
        }
        return ans;
    }

    Reason for rejection:
    - Very slow (O(n × x))
    - Will TLE for large numbers
    */
};

File: Daily-Streak/Day-04_Four-Divisors/Solution.py
class Solution:
    def sumFourDivisors(self, nums: List[int]) -> int:
        answer = 0

        # ✅ OPTIMAL APPROACH (ACTIVE CODE)
        for x in nums:
            count = 0
            total = 0
            i = 1

            while i * i <= x:
                if x % i == 0:
                    d1 = i
                    d2 = x // i

                    count += 1
                    total += d1

                    if d1 != d2:
                        count += 1
                        total += d2

                    # Early stopping
                    if count > 4:
                        break
                i += 1

            if count == 4:
                answer += total

        return answer

        """
        ❌ BRUTE FORCE APPROACH (COMMENTED)

        ans = 0
        for x in nums:
            divisors = []
            for i in range(1, x + 1):
                if x % i == 0:
                    divisors.append(i)
            if len(divisors) == 4:
                ans += sum(divisors)
        return ans

        Reason:
        - Extremely slow
        - Not scalable
        """
